libraries/llvm/io.c

```diff
@@ -27,159 +27,128 @@ void c_io_println_String(String text) {
+            return UV_FS_O_RDONLY;
+        case 1: // WriteOnly()
+            return UV_FS_O_WRONLY | UV_FS_O_CREAT | UV_FS_O_TRUNC;
+        case 2: // AppendOnly()
+            return UV_FS_O_WRONLY | UV_FS_O_CREAT | UV_FS_O_APPEND;
+        case 3: // ReadWrite()
+            return UV_FS_O_RDWR | UV_FS_O_CREAT | UV_FS_O_TRUNC;
+        case 4: // ReadAppend()
+            return UV_FS_O_RDWR | UV_FS_O_CREAT | UV_FS_O_APPEND;
+        case 5: // AppendExclusive()
+            return UV_FS_O_WRONLY | UV_FS_O_CREAT | UV_FS_O_APPEND | UV_FS_O_EXCL;
+        case 6: // ReadAppendExclusive()
+            return UV_FS_O_RDWR | UV_FS_O_CREAT | UV_FS_O_APPEND | UV_FS_O_EXCL;
+        case 7: // AppendSync()
+            return UV_FS_O_WRONLY | UV_FS_O_CREAT | UV_FS_O_APPEND | UV_FS_O_SYNC;
+        case 8: // ReadAppendSync()
+            return UV_FS_O_RDWR | UV_FS_O_CREAT | UV_FS_O_APPEND | UV_FS_O_SYNC;
+        case 9: // ReadSync()
+            return UV_FS_O_RDONLY | UV_FS_O_SYNC;
+        case 10: // ReadWriteSync()
+            return UV_FS_O_RDWR | UV_FS_O_SYNC;
+        case 11: // WriteExclusive()
+            return UV_FS_O_WRONLY | UV_FS_O_CREAT | UV_FS_O_TRUNC | UV_FS_O_EXCL;
+        case 12: // ReadWriteExclusive()
+            return UV_FS_O_RDWR | UV_FS_O_CREAT | UV_FS_O_TRUNC | UV_FS_O_EXCL;
+        default:
+            // Invalid tag value
+            return -1;
     }
 }
 
-void c_promise_resolve(struct Pos promise, struct Pos value) {
-    Promise* p = (Promise*)promise.obj;
+void c_fs_open(struct Pos path, struct Pos mode, Stack stack) {
 
-    switch (p->state) {
-        case UNRESOLVED:
-            p->state = RESOLVED;
-            p->payload.pos = value; // Store value in payload
-            break;
-        case RESOLVED:
-            fprintf(stderr, "ERROR: Promise already resolved\n");
-            exit(1);
-        case AWAITED: {
-            Listeners* current = p->payload.listeners;
-            p->state = RESOLVED;
-            p->payload.pos = value;
-
-             // Call each listeners
-            while (current != NULL) {
-                sharePositive(value);
-                run_Pos(current->listener, value);
-                Listeners* temp = current;
-                current = current->next;
-                free(temp);
-            }
-            break;
-        }
-    }
-    // do we need to erase promise now? Is it shared before?
-    erasePositive(promise);
+    // Convert the Effekt String to a 0-terminated string
+    char* path_str = c_buffer_as_null_terminated_string(path);
+    erasePositive((struct Pos) path);
+
+    // Convert the Effekt String representing the opening mode to libuv flags
+    int flags = modeFlags(mode);
+    erasePositive((struct Pos) mode);
+
+    uv_fs_t* request = malloc(sizeof(uv_fs_t));
+    request->data = stack;
+
+    uv_fs_open(uv_default_loop(), request, path_str, flags, 0666, c_resume_int_fs);
+    // TODO report result (UV_EINVAL)
```

Comment:
Do we need to free stuff when this errors? This probably means the continuation is never called and also needs to be ref-count-decremented.